python/vowpalwabbit/pyvw.py

```diff
@@ -80,6 +81,78 @@ def __str__(self):
         else:
             return ''
 
+class Learner:
+    def __init__(self, vwCppBridge):
+        self.vwCppBridge = vwCppBridge
+
+    def learn(self, ec):
+        self.vwCppBridge.call_base_learner(ec, True)
+
+    def predict(self, ec):
+        self.vwCppBridge.call_base_learner(ec, False)
+
+    def multi_learn(self, examples, offset = 0, id = 0):
+        self.vwCppBridge.call_multi_learner(examples, True)
+
+    def multi_predict(self, examples, offset = 0, id = 0):
+        self.vwCppBridge.call_multi_learner(examples, False)
+
+class ModelIO:
+    def __init__(self, vwCppBridge):
+        self.vwCppBridge = vwCppBridge
+
+    # (char* data, size_t len, const char* read_message, bool read, std::stringstream& msg, bool text)
+    def read_write(self):
+        print("read and writing")
+
+# compatible with Python 2 *and* 3
+ABC = abc.ABCMeta('ABC', (object,), {'__slots__': ()})
+
+def no_impl(method):
+  method.no_impl = True
+  return method
+
+# End user (pyvw library consumer) will have to create a class that
+# inherits from Copperhead and implements the custom _predict() and
+# _learn(). Optionally the user can also implement _finish_example()
+# if needed. See test_pyreduction.py for examples
+"""Copperhead class"""
+class Copperhead(ABC):
```

Comment:
I find the name `Copperhead` kind of confusing and not indicative of what this class does

Reply:
renamed